File: statistics.ml
(*
 * Statistics Calculator - OCaml Implementation (Functional Paradigm)
 * 
 * This program calculates mean, median, and mode of a list of integers
 * using a functional approach with immutable data structures.
 *)

(* Calculate the mean (average) of integers *)
let calculate_mean lst =
  if lst = [] then 0.0
  else
    let sum = List.fold_left (fun acc x -> acc + x) 0 lst in
    float_of_int sum /. float_of_int (List.length lst)

(* Calculate the median (middle value when sorted) *)
let calculate_median lst =
  if lst = [] then 0.0
  else
    let sorted = List.sort compare lst in
    let len = List.length sorted in
    if len mod 2 = 0 then
      (* Even number of elements: average of two middle values *)
      let mid1 = List.nth sorted (len / 2 - 1) in
      let mid2 = List.nth sorted (len / 2) in
      (float_of_int mid1 +. float_of_int mid2) /. 2.0
    else
      (* Odd number of elements: middle value *)
      float_of_int (List.nth sorted (len / 2))

(* Count frequency of each element in the list *)
let count_frequencies lst =
  List.fold_left
    (fun acc x ->
      if List.mem_assoc x acc then
        List.map (fun (k, v) -> if k = x then (k, v + 1) else (k, v)) acc
      else
        (x, 1) :: acc)
    [] lst

(* Find the maximum frequency *)
let max_frequency freq_list =
  if freq_list = [] then 0
  else
    List.fold_left (fun max_val (_, freq) -> if freq > max_val then freq else max_val) 0 freq_list

(* Calculate the mode (most frequently occurring value(s)) *)
let calculate_mode lst =
  if lst = [] then []
  else
    let freq_list = count_frequencies lst in
    let max_freq = max_frequency freq_list in
    List.fold_left
      (fun acc (value, freq) ->
        if freq = max_freq then value :: acc else acc)
      [] freq_list
    |> List.sort compare

(* Format list as string for printing *)
let list_to_string lst =
  "[" ^ String.concat ", " (List.map string_of_int lst) ^ "]"

(* Print statistics *)
let print_statistics lst =
  print_endline "\n=== Statistics Results ===";
  print_string "Data: ";
  print_endline (list_to_string lst);
  print_newline ();
  
  (* Calculate and print mean *)
  let mean = calculate_mean lst in
  Printf.printf "Mean: %.2f\n" mean;
  
  (* Calculate and print median *)
  let median = calculate_median lst in
  Printf.printf "Median: %.2f\n" median;
  
  (* Calculate and print mode *)
  let modes = calculate_mode lst in
  print_string "Mode: ";
  if modes = [] then
    print_endline "No mode"
  else
    let mode_str = String.concat ", " (List.map string_of_int modes) in
    Printf.printf "%s\n" mode_str;
  
  print_endline "========================\n"

(* Main function *)
let main () =
  print_endline "Statistics Calculator - OCaml Implementation";
  print_endline "=============================================";
  
  (* Example 1: Basic example *)
  let data1 = [1; 2; 3; 4; 5] in
  print_statistics data1;
  
  (* Example 2: With duplicates *)
  let data2 = [1; 2; 2; 3; 3; 3; 4; 4; 5] in
  print_statistics data2;
  
  (* Example 3: Even number of elements *)
  let data3 = [10; 20; 30; 40] in
  print_statistics data3;
  
  (* Example 4: Single mode *)
  let data4 = [5; 5; 5; 2; 2; 1] in
  print_statistics data4;
  
  (* Example 5: Multiple modes *)
  let data5 = [1; 1; 2; 2; 3; 3] in
  print_statistics data5

(* Run the program *)
let () = main ()

File: statistics.py
"""
Statistics Calculator - Python Implementation (Object-Oriented Paradigm)

This program calculates mean, median, and mode of a list of integers
using an object-oriented approach with a StatisticsCalculator class.
"""

from collections import Counter
from typing import List


class StatisticsCalculator:
    """
    A class to calculate basic statistics (mean, median, mode) 
    on a list of integers.
    """
    
    def __init__(self, data: List[int]):
        """
        Initialize the StatisticsCalculator with a list of integers.
        
        Args:
            data: List of integers to calculate statistics on
        """
        self.data = data
    
    def calculate_mean(self) -> float:
        """
        Calculate the mean (average) of the integers.
        
        Returns:
            The mean value as a float. Returns 0.0 if the list is empty.
        """
        if not self.data:
            return 0.0
        
        return sum(self.data) / len(self.data)
    
    def calculate_median(self) -> float:
        """
        Calculate the median (middle value when sorted) of the integers.
        
        Returns:
            The median value as a float. Returns 0.0 if the list is empty.
        """
        if not self.data:
            return 0.0
        
        sorted_data = sorted(self.data)
        length = len(sorted_data)
        
        if length % 2 == 0:
            # Even number of elements: average of two middle values
            mid1 = sorted_data[length // 2 - 1]
            mid2 = sorted_data[length // 2]
            return (mid1 + mid2) / 2.0
        else:
            # Odd number of elements: middle value
            return float(sorted_data[length // 2])
    
    def calculate_mode(self) -> List[int]:
        """
        Calculate the mode (most frequently occurring value(s)) of the integers.
        
        Returns:
            A list of mode values. Returns an empty list if the input is empty.
            If multiple values have the same maximum frequency, all are returned.
        """
        if not self.data:
            return []
        
        # Count frequency of each element
        frequency_counter = Counter(self.data)
        
        # Find the maximum frequency
        max_frequency = max(frequency_counter.values())
        
        # Find all values with the maximum frequency
        modes = [value for value, freq in frequency_counter.items() 
                if freq == max_frequency]
        
        # Sort modes for consistent output
        return sorted(modes)
    
    def print_statistics(self) -> None:
        """
        Calculate and print all statistics (mean, median, mode).
        """
        print("\n=== Statistics Results ===")
        print(f"Data: {self.data}\n")
        
        # Calculate and print mean
        mean = self.calculate_mean()
        print(f"Mean: {mean:.2f}")
        
        # Calculate and print median
        median = self.calculate_median()
        print(f"Median: {median:.2f}")
        
        # Calculate and print mode
        modes = self.calculate_mode()
        print("Mode: ", end="")
        if not modes:
            print("No mode")
        else:
            mode_str = ", ".join(str(m) for m in modes)
            print(mode_str)
        
        print("========================\n")
    
    def get_all_statistics(self) -> dict:
        """
        Calculate and return all statistics as a dictionary.
        
        Returns:
            A dictionary containing 'mean', 'median', and 'mode' keys.
        """
        return {
            'mean': self.calculate_mean(),
            'median': self.calculate_median(),
            'mode': self.calculate_mode()
        }


def main():
    """
    Main function demonstrating the StatisticsCalculator class.
    """
    print("Statistics Calculator - Python Implementation")
    print("==============================================")
    
    # Example 1: Basic example
    calc1 = StatisticsCalculator([1, 2, 3, 4, 5])
    calc1.print_statistics()
    
    # Example 2: With duplicates
    calc2 = StatisticsCalculator([1, 2, 2, 3, 3, 3, 4, 4, 5])
    calc2.print_statistics()
    
    # Example 3: Even number of elements
    calc3 = StatisticsCalculator([10, 20, 30, 40])
    calc3.print_statistics()
    
    # Example 4: Single mode
    calc4 = StatisticsCalculator([5, 5, 5, 2, 2, 1])
    calc4.print_statistics()
    
    # Example 5: Multiple modes
    calc5 = StatisticsCalculator([1, 1, 2, 2, 3, 3])
    calc5.print_statistics()
    
    # Example 6: Using get_all_statistics method
    calc6 = StatisticsCalculator([7, 8, 9, 9, 10])
    stats = calc6.get_all_statistics()
    print("Example using get_all_statistics():")
    print(f"Statistics: {stats}\n")


if __name__ == "__main__":
    main()

